jetty-core/jetty-websocket/jetty-websocket-jetty-common/src/main/java/org/eclipse/jetty/websocket/common/JettyWebSocketFrameHandlerFactory.java

```diff
@@ -167,76 +174,127 @@ private JettyWebSocketFrameHandlerMetadata createListenerMetadata(Class<?> endpo
             metadata.setBinaryHandle(ByteBufferMessageSink.class, binaryHandle, binaryMethod);
         }
 
         Method errorMethod = findMethod(endpointClass, "onWebSocketError", Throwable.class);
         if (errorMethod != null)
         {
-            MethodHandle errorHandle = toMethodHandle(lookup, errorMethod);
+            MethodHolder errorHandle = toMethodHolder(lookup, errorMethod);
             metadata.setErrorHandle(errorHandle, errorMethod);
         }
 
-        Method closeMethod = findMethod(endpointClass, "onWebSocketClose", int.class, String.class);
+        Method deprecatedCloseMethod = findMethod(endpointClass, "onWebSocketClose", int.class, String.class);
+        Method closeMethod = findMethod(endpointClass, "onWebSocketClose", int.class, String.class, Callback.class);
         if (closeMethod != null)
         {
-            MethodHandle closeHandle = toMethodHandle(lookup, closeMethod);
+            if (deprecatedCloseMethod != null)
+                throw new InvalidWebSocketException("Cannot use two versions of onWebSocketClose");
+
+            MethodHolder closeHandle = toMethodHolder(lookup, closeMethod);
             metadata.setCloseHandle(closeHandle, closeMethod);
         }
+        else if (deprecatedCloseMethod != null)
+        {
+            MethodHandle deprecatedCloseHandle = toMethodHandle(lookup, deprecatedCloseMethod);
+            deprecatedCloseHandle = MethodHandles.dropArguments(deprecatedCloseHandle, 3, Callback.class);
+            MethodHolder closeHandle = new CallbackCompletingCloseHolder(MethodHolder.from(deprecatedCloseHandle));
+            metadata.setCloseHandle(closeHandle, deprecatedCloseMethod);
+        }
 
         return metadata;
     }
 
+    /**
+     * This class wraps a {@link MethodHolder} and automatically completes the callback in the signature when the
+     * call to the wrapped {@link MethodHolder} returns.
+     */
+    private static class CallbackCompletingCloseHolder extends MethodHolder.Wrapper
+    {
+        public CallbackCompletingCloseHolder(MethodHolder methodHolder)
+        {
+            super(methodHolder);
+        }
+
+        @Override
+        public Object invoke(Object... args)
+        {
+            Callback callback = (Callback)Arrays.stream(args)
+                .filter(o -> o instanceof Callback)
+                .findFirst()
+                .orElseThrow(IllegalArgumentException::new);
+
+            try
+            {
+                Object value = getWrapped().invoke(args);
+                callback.succeed();
+                return value;
+            }
+            catch (Throwable t)
+            {
+                callback.fail(t);
+            }
+            return null;
+        }
+
+        @Override
+        public MethodHolder bindTo(Object arg, int idx)
+        {
+            return new CallbackCompletingCloseHolder(getWrapped().bindTo(arg, idx));
```

Comment:
```suggestion
            return new CallbackCompletingCloseHolder(super.bindTo(arg, idx));
```

There is another bindTo() method to override.